testing/endtoend/e2e_crash_test.go

```diff
@@ -0,0 +1,303 @@
+					continue
+				}
+				// We expect the honest party to have confirmed it.
+				if !honestConfirmed {
+					t.Fatal("Evil party confirmed the assertion by challenge win")
+				}
+				confirmed = true
+				break
+			}
+			time.Sleep(500 * time.Millisecond) // Don't spam the backend.
+		}
+		// Once the honest, claimed assertion in the challenge is confirmed by time, we
+		// then continue the test.
+		t.Log("Assertion was confirmed by time")
+	})
+	// This test ensures that an honest validator can crash after a challenge has completed, can resync
+	// the completed challenge and continue playing the game until all essential edges are confirmed.
+	// This is to ensure that even if a challenge is completed, we can still resync it and continue
+	// playing for the sake of refunding honest stakes.
+	t.Run(
+		"crashes once challenged assertion is confirmed and restarts to confirm essential edges",
+		func(t *testing.T) {
+			// We start by canceling the honest context, and then wait a few seconds, and expect the honest
+			// validator to resync the challenge that just completed to continue playing the game
+			// until all essential root edges are confirmed by time.
+			honestCancel()
+			t.Log("Honest context has been canceled after challenge win")
+
+			// We then restart the honest validator after a few seconds of wait time.
+			time.Sleep(time.Second * 30)
+
+			honestCtx, honestCancel = context.WithCancel(context.Background())
+			honestChain = setupAssertionChain(t, honestCtx, bk.Client(), rollupAddr.Rollup, txOpts)
+			honestManager, err = cm.NewChallengeStack(honestChain, honestStateManager, honestOpts...)
+			require.NoError(t, err)
+
+			honestManager.Start(honestCtx)
+
+			t.Log("Restarted honest validator to continue playing game after challenge has finished")
+
+			// We then expect that all essential root edges created by the honest validator are confirmed by time.
+			// Scrape all the honest edges onchain (the ones made by the honest address).
+			// Check if the edges that have claim id != None are confirmed (those are essential root edges)
+			// and also check one step edges from honest party are confirmed.
+			honestEssentialRootIds := make(map[common.Hash]bool, 0)
+			chainId, err := bk.Client().ChainID(neutralCtx)
+			require.NoError(t, err)
+			it, err := cmBindings.FilterEdgeAdded(nil, nil, nil, nil)
+			require.NoError(t, err)
+			for it.Next() {
+				txHash := it.Event.Raw.TxHash
+				tx, _, err := bk.Client().TransactionByHash(neutralCtx, txHash)
+				require.NoError(t, err)
+				sender, err := gethtypes.Sender(gethtypes.NewCancunSigner(chainId), tx)
+				require.NoError(t, err)
+				if sender != txOpts.From {
+					continue
+				}
+				// Skip edges that are not essential roots.
+				if it.Event.ClaimId == (common.Hash{}) {
+					continue
+				}
+				honestEssentialRootIds[it.Event.EdgeId] = false
+			}
+			// Wait until all of the honest essential root ids are confirmed.
+			confirmedCount := 0
```

Comment:
OPTIONAL: Is the only way for this test to fail through "timeout?"
Is there anything we could do to make it fail faster if the underlying code is broken? For example, would it make sense to ensure that we are working through the backlog of missing confirmations at least once per 2 seconds or once per some number of L1 blocks, or something?

Reply:
When exactly each essential edge will be confirmed is hard to tell, as the moves are all fairly concurrent and there is a lot going on within the test. It also depends on when each tx gets included onchain, and depends on config params per test. Not sure how else we could structure this one at the moment...maybe we could error if no confirmation has occurred in 1 challenge period